### PR TITLE
new internal API for sampling

### PR DESCRIPTION
**Pull request by Tortar**
This will solve #837 when finished

### CODE REVIEW

**Comment by Tortar** [comment]
This will actually be solved when I manage to find the time to finish the work in https://github.com/Tortar/IteratorSampling.jl

hopefully soon

**Comment by Tortar** [comment]
Everything works now, so I think it should good to go

I needed to add some more Pkg tricks though because I noticed that if you didn't have internet connection the solution in #955 didn't work, I verified that instead works like this. This is only temporary, when we release v6 we will rapidly delete all of this.

I made our sampling depending from https://github.com/Tortar/IteratorSampling.jl where the sampling process is much more well tested than how it was in our library, so I guess it's okay to depend on it :-)

Anyway there we have also fast sampling for multiple items, but here I didn't include it because it requires some more work

**Comment by Tortar** [comment]
actually there are some ensemble tests which didn't fail locally, not sure why at the moment

**Comment by Tortar** [comment]
mmmmh, they seem actually totally unrelated

File: src/simulations/paramscan.jl [hunk 1]
```diff
@@ -101,6 +101,11 @@ function paramscan(
 
     progress = ProgressMeter.Progress(length(combs); enabled = showprogress)
     mapfun = parallel ? pmap : map
+    println(combs)
+    println(mapfun)
+    println(output_params)
+    println(initialize)
+    println(kwargs)
```
Comment (Datseris) [inline src/simulations/paramscan.jl:108]: Mistake?

File: src/core/space_interaction_API.jl [hunk 2]
```diff
@@ -380,69 +377,13 @@ is expensive since in this case the allocating version can be more performant.
 function random_nearby_position(pos, model, r=1, f = nothing, alloc = false; kwargs...)
     iter = nearby_positions(pos, model, r; kwargs...)
     if isnothing(f)
-        return resorvoir_sampling_single(iter, model)
+        return itsample(abmrng(model), iter)
```
Comment (Datseris) [inline src/core/space_interaction_API.jl:380]: Does `itsample` come from your package? I would say, preface the functions that come from your package with `IteratorSampling.function_name`.

File: src/core/space_interaction_API.jl [hunk 3]
```diff
@@ -304,8 +304,7 @@ nearby_agents(a, model, r = 1; kwargs...) =
 
 """
     random_nearby_id(agent, model::ABM, r = 1, f = nothing, alloc = false; kwargs...) → id
-Return the `id` of a random agent near the position of the given `agent` using an optimized
-algorithm from [Reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling#An_optimal_algorithm).
+Return the `id` of a random agent near the position of the given `agent`.
```
Comment (Datseris) [inline src/core/space_interaction_API.jl:307]: It's fine to not say too many details, but i think it is still improtant to say that this function does not require collecting all the agents and randomly sampling one.

**Review by Datseris — CHANGES_REQUESTED** [review]
Have you checked for performance regressions? The introduction of `Iteartors.Filter` e.g. could add some.

File: src/simulations/paramscan.jl (hunk 1 above)
Comment (Tortar) [reply src/simulations/paramscan.jl:108]: yes, forgot about removing them

File: src/core/space_interaction_API.jl (hunk 2 above)
Comment (Tortar) [reply src/core/space_interaction_API.jl:380]: yes, good idea

File: src/core/space_interaction_API.jl (hunk 3 above)
Comment (Tortar) [reply src/core/space_interaction_API.jl:307]: reintroduced in a similar form

**Comment by Tortar** [comment]
checked briefly with this code

```julia
using Agents, BenchmarkTools

@agent A GridAgent{2} begin
end

agent_step!(a,m) = nothing

function aperiodic_model(; griddims = (100, 100))
    space = GridSpace(griddims, periodic = false)
    model = ABM(A, space; agent_step!)
    fill_space!(model)
    return model
end

function periodic_model(; griddims = (100, 100))
    space = GridSpace(griddims, periodic = true)
    model = ABM(A, space; agent_step!)
    fill_space!(model)
    return model
end

@btime random_nearby_id($(1,1), model, 10) setup=(model=periodic_model());
@btime random_nearby_id($(50,50), model, 10) setup=(model=periodic_model());
@btime random_nearby_id($(1,1), model, 1) setup=(model=aperiodic_model());
@btime random_nearby_id($(50,50), model, 1) setup=(model=aperiodic_model());
```

no perf regression, same perf, in general I expect the same for all the other function apart from a little improvement where I introduced the new `Iterators.filter` in `random_empty_pos_in_offsets` because it is a fallback when there is a small number of empty cells, in this case using the other methodology  (`sampling_with_condition_single` function) is bad for performance.

So I think this PR should be good to go as it is. Let's see in the future if we want to support optimized multi sampling procedures or maybe give to the users some advice on how to perform them by themselves. Maybe we can discuss that in our call too

**Comment by Datseris** [comment]
yes, this is good from me as well, but all tests fail!

**Comment by Tortar** [comment]
I know but the tests that fail seem spurious, this doesn't happen locally, it happens also in #946 and #957

**Comment by Tortar** [comment]
If you can try it out too locally it would be better to conclude about that something is strange just on github

**Review by Datseris — APPROVED** [review]
If tests pass locally I trust you. We would need to figure out how to solve this before v6. (if tests fail on `main` after merging this, we should also open an issue to keep track of things.

**Comment by Tortar** [comment]
tests passed when you merged, so I think it alright

**Pull request merged**